File: haskell/space-age/src/SpaceAge.hs
module SpaceAge (Planet(..), ageOn) where

data Planet = Mercury | Venus | Earth | Mars |
              Jupiter | Saturn | Uranus | Neptune

earthYear = 31557600 -- seconds

-- Returns the number of Earth years equivalent to one year on a given planet.
orbitalPeriod :: Planet -> Float
orbitalPeriod Earth   = 1
orbitalPeriod Mercury = 0.2408467
orbitalPeriod Venus   = 0.61519726
orbitalPeriod Mars    = 1.8808158
orbitalPeriod Jupiter = 11.862615
orbitalPeriod Saturn  = 29.447498
orbitalPeriod Uranus  = 84.016846
orbitalPeriod Neptune = 164.79132

oneYearInSeconds :: Planet -> Float
oneYearInSeconds planet = earthYear * (orbitalPeriod planet)

ageOn :: Planet -> Float -> Float
ageOn planet seconds = seconds / (oneYearInSeconds planet)

File: elixir/space-age/space_age.exs
defmodule SpaceAge do
  @type planet :: :mercury | :venus | :earth | :mars | :jupiter
                | :saturn | :uranus | :neptune

  @earth_year 31557600 # seconds

  # The number of Earth years equivalent to one year on a given planet.
  @orbital_period %{
    earth: 1,
    mercury: 0.2408467,
    venus: 0.61519726,
    mars: 1.8808158,
    jupiter: 11.862615,
    saturn: 29.447498,
    uranus: 84.016846,
    neptune: 164.79132
  }

  @doc """
  Returns the number of seconds in a year on a given planet.
  """
  @spec one_year_in_seconds_on(planet) :: float
  def one_year_in_seconds_on(planet) do
    @earth_year * @orbital_period[planet]
  end

  @doc """
  Return the number of years a person that has lived for 'seconds' seconds is
  aged on 'planet'.
  """
  @spec age_on(planet, pos_integer) :: float
  def age_on(planet, seconds) do
    seconds / one_year_in_seconds_on planet
  end
end

